File: Sources/OpenZCineCore/ISOPickerPolicy.swift
/// ISO picker layout and recording-lock rules for Nikon ZR dual-base sensitivity.
///
/// R3D NE exposes separate low/high base ISO circuits in the UI; other codecs keep dual-base
/// hardware but auto-switch the circuit, so the operator sees one drum with native-base markers.
public enum ISOPickerPolicy: Sendable {
    /// Low-base ISO steps (200–3200).
    public static let lowBaseOptions: [String] = [
        "200", "250", "320", "400", "500", "640", "800", "1000", "1250", "1600",
        "2000", "2500", "3200",
    ]

    /// High-base ISO steps (1600–25600).
    public static let highBaseOptions: [String] = [
        "1600", "2000", "2500", "3200", "4000", "5000", "6400", "8000", "10000",
        "12800", "16000", "20000", "25600",
    ]

    /// Native low-base ISO flagged in the drum.
    public static let lowBaseMarker = "800"

    /// Native high-base ISO flagged in the drum.
    public static let highBaseMarker = "6400"

    /// Single-drum ISO steps for codecs that auto-switch base circuits (union of both ranges).
    public static let unifiedOptions: [String] = {
        var seen = Set<String>()
        var merged: [String] = []
        for value in lowBaseOptions + highBaseOptions where seen.insert(value).inserted {
            merged.append(value)
        }
        return merged
    }()

    /// Whether the active codec is R3D NE (accepts raw or shortened camera labels).
    public static func isR3DNECodec(_ codec: String) -> Bool {
        MonitorTextFormat.codecShortLabel(codec) == "R3D NE"
    }

    /// R3D NE keeps separate LOW/HIGH base drums; every other codec uses a unified drum.
    public static func showsDualBaseCircuits(codec: String) -> Bool {
        isR3DNECodec(codec)
    }

    /// Non-R3D NE codecs use Auto On/Off tabs instead of dual-base circuits.
    public static func showsAutoISOControl(codec: String) -> Bool {
        !showsDualBaseCircuits(codec: codec)
    }

    /// Exposure modes where the body typically owns ISO (auto ISO). Manual ISO is free in `M`.
    public static func isAutoISOActive(exposureMode: String?) -> Bool {
        guard let mode = exposureMode?.trimmingCharacters(in: .whitespacesAndNewlines), !mode.isEmpty
        else { return false }
        switch mode {
        case "Auto", "P", "A", "S":
            return true
        default:
            // U1–U3 and M are treated as operator-controlled ISO unless the body reports Auto.
            return mode.localizedCaseInsensitiveContains("auto")
        }
    }

    /// Exposure-mode label that enables auto ISO (full Auto program).
    public static let autoISOOnExposureMode = "Auto"

    /// Exposure-mode label that releases ISO for the drum (manual movie exposure).
    public static let autoISOOffExposureMode = "M"

    /// ISO cannot be changed while recording in R3D NE; other codecs allow mid-roll ISO changes.
    public static func blocksISOChangeWhileRecording(codec: String, isRecording: Bool) -> Bool {
        isR3DNECodec(codec) && isRecording
    }

    /// Star markers for the active layout. Dual-base marks one native base per circuit tab;
    /// unified / Auto ISO marks both native bases on the single drum.
    public static func markedValues(codec: String, modeIndex: Int) -> Set<String> {
        if showsDualBaseCircuits(codec: codec) {
            guard modeIndex == 0 || modeIndex == 1 else { return [] }
            return [modeIndex == 0 ? lowBaseMarker : highBaseMarker]
        }
        return [lowBaseMarker, highBaseMarker]
    }

    /// Subtitle shown under the ISO picker header.
    public static func pickerSubtitle(codec: String) -> String {
        if showsDualBaseCircuits(codec: codec) { return "Sensitivity · dual base" }
        if showsAutoISOControl(codec: codec) { return "Sensitivity · auto / manual" }
        return "Sensitivity"
    }

    /// Mode tabs for the ISO picker.
    /// R3D NE → Low/High base; other codecs → Auto On / Auto Off; never empty for ZR movie codecs.
    public static func pickerModes(codec: String) -> [ISOPickerMode] {
        if showsDualBaseCircuits(codec: codec) {
            return [
                ISOPickerMode(
                    title: "Low Base",
                    detail: "\(lowBaseMarker) · 200-3200",
                    options: lowBaseOptions,
                    base: lowBaseMarker
                ),
                ISOPickerMode(
                    title: "High Base",
                    detail: "\(highBaseMarker) · 1600-25600",
                    options: highBaseOptions,
                    base: highBaseMarker
                ),
            ]
        }
        if showsAutoISOControl(codec: codec) {
            return [
                ISOPickerMode(
                    title: "Auto On",
                    detail: "Camera controls ISO",
                    options: unifiedOptions,
                    base: lowBaseMarker,
                    activatesAutoISO: true
                ),
                ISOPickerMode(
                    title: "Auto Off",
                    detail: "Manual ISO",
                    options: unifiedOptions,
                    base: lowBaseMarker,
                    activatesAutoISO: false
                ),
            ]
        }
        return []
    }

    /// Options for the active ISO layout and mode tab.
    public static func options(codec: String, modeIndex: Int) -> [String] {
        let modes = pickerModes(codec: codec)
        guard !modes.isEmpty, modes.indices.contains(modeIndex) else { return unifiedOptions }
        return modes[modeIndex].options
    }

    /// Active mode tab for Auto ISO codecs: 0 = Auto On, 1 = Auto Off.
    public static func autoISOModeIndex(exposureMode: String?) -> Int {
        isAutoISOActive(exposureMode: exposureMode) ? 0 : 1
    }
}

/// One segmented mode under the ISO picker (LOW/HIGH base or Auto On/Off).
public struct ISOPickerMode: Equatable, Sendable {
    public let title: String
    public let detail: String?
    public let options: [String]
    public let base: String
    /// When non-nil, selecting this tab should switch camera auto-ISO on (`true`) or off (`false`).
    public let activatesAutoISO: Bool?

    public init(
        title: String,
        detail: String? = nil,
        options: [String],
        base: String,
        activatesAutoISO: Bool? = nil
    ) {
        self.title = title
        self.detail = detail
        self.options = options
        self.base = base
        self.activatesAutoISO = activatesAutoISO
    }
}

File: Apps/Android/app/src/main/kotlin/com/opencapture/openzcine/IsoPickerPolicy.kt
package com.opencapture.openzcine

/**
 * ISO picker layout and recording-lock rules for Nikon ZR dual-base sensitivity.
 *
 * 1:1 with `Sources/OpenZCineCore/ISOPickerPolicy.swift`. R3D NE exposes separate
 * low/high base ISO circuits; other codecs use Auto On/Off plus a unified drum.
 */
internal object IsoPickerPolicy {
    /** Low-base ISO steps (200–3200). */
    val lowBaseOptions: List<String> =
        listOf(
            "200",
            "250",
            "320",
            "400",
            "500",
            "640",
            "800",
            "1000",
            "1250",
            "1600",
            "2000",
            "2500",
            "3200",
        )

    /** High-base ISO steps (1600–25600). */
    val highBaseOptions: List<String> =
        listOf(
            "1600",
            "2000",
            "2500",
            "3200",
            "4000",
            "5000",
            "6400",
            "8000",
            "10000",
            "12800",
            "16000",
            "20000",
            "25600",
        )

    /** Native low-base ISO flagged in the drum. */
    const val LOW_BASE_MARKER: String = "800"

    /** Native high-base ISO flagged in the drum. */
    const val HIGH_BASE_MARKER: String = "6400"

    /** Exposure mode that enables camera-managed ISO. */
    const val AUTO_ISO_ON_EXPOSURE_MODE: String = "Auto"

    /** Exposure mode that releases ISO for the drum. */
    const val AUTO_ISO_OFF_EXPOSURE_MODE: String = "M"

    /** Single-drum ISO steps for codecs that auto-switch base circuits. */
    val unifiedOptions: List<String> =
        (lowBaseOptions + highBaseOptions).distinct()

    /** Whether the active codec is R3D NE (raw or shortened camera labels). */
    fun isR3DNECodec(codec: String): Boolean {
        val short =
            codec
                .trim()
                .let { raw ->
                    when {
                        raw.contains("R3D", ignoreCase = true) &&
                            raw.contains("NE", ignoreCase = true) -> "R3D NE"
                        else -> raw
                    }
                }
        return short == "R3D NE"
    }

    /** R3D NE keeps separate LOW/HIGH base drums. */
    fun showsDualBaseCircuits(codec: String): Boolean = isR3DNECodec(codec)

    /** Non-R3D NE codecs use Auto On/Off tabs. */
    fun showsAutoISOControl(codec: String): Boolean = !showsDualBaseCircuits(codec)

    /** Exposure modes where the body typically owns ISO. */
    fun isAutoISOActive(exposureMode: String?): Boolean {
        val mode = exposureMode?.trim().orEmpty()
        if (mode.isEmpty()) return false
        return when (mode) {
            "Auto", "P", "A", "S" -> true
            else -> mode.contains("auto", ignoreCase = true)
        }
    }

    /** Active Auto tab: 0 = Auto On, 1 = Auto Off. */
    fun autoISOModeIndex(exposureMode: String?): Int =
        if (isAutoISOActive(exposureMode)) 0 else 1

    /** ISO cannot be changed while recording in R3D NE. */
    fun blocksISOChangeWhileRecording(codec: String, isRecording: Boolean): Boolean =
        isR3DNECodec(codec) && isRecording

    /** Star markers for the active layout. */
    fun markedValues(codec: String, modeIndex: Int): Set<String> =
        if (showsDualBaseCircuits(codec)) {
            when (modeIndex) {
                0 -> setOf(LOW_BASE_MARKER)
                1 -> setOf(HIGH_BASE_MARKER)
                else -> emptySet()
            }
        } else {
            setOf(LOW_BASE_MARKER, HIGH_BASE_MARKER)
        }

    /** Subtitle shown under the ISO picker header. */
    fun pickerSubtitle(codec: String): String =
        when {
            showsDualBaseCircuits(codec) -> "Sensitivity · dual base"
            showsAutoISOControl(codec) -> "Sensitivity · auto / manual"
            else -> "Sensitivity"
        }

    /** Mode tabs for the ISO picker. */
    fun pickerModes(codec: String): List<IsoPickerMode> =
        when {
            showsDualBaseCircuits(codec) ->
                listOf(
                    IsoPickerMode(
                        title = "Low Base",
                        detail = "$LOW_BASE_MARKER · 200-3200",
                        options = lowBaseOptions,
                        base = LOW_BASE_MARKER,
                    ),
                    IsoPickerMode(
                        title = "High Base",
                        detail = "$HIGH_BASE_MARKER · 1600-25600",
                        options = highBaseOptions,
                        base = HIGH_BASE_MARKER,
                    ),
                )
            showsAutoISOControl(codec) ->
                listOf(
                    IsoPickerMode(
                        title = "Auto On",
                        detail = "Camera controls ISO",
                        options = unifiedOptions,
                        base = LOW_BASE_MARKER,
                        activatesAutoISO = true,
                    ),
                    IsoPickerMode(
                        title = "Auto Off",
                        detail = "Manual ISO",
                        options = unifiedOptions,
                        base = LOW_BASE_MARKER,
                        activatesAutoISO = false,
                    ),
                )
            else -> emptyList()
        }

    /** Options for the active ISO layout and mode tab. */
    fun options(codec: String, modeIndex: Int): List<String> {
        val modes = pickerModes(codec)
        return if (modes.isEmpty() || modeIndex !in modes.indices) {
            unifiedOptions
        } else {
            modes[modeIndex].options
        }
    }
}

/** One segmented mode under the ISO picker (LOW/HIGH base or Auto On/Off). */
internal data class IsoPickerMode(
    val title: String,
    val detail: String?,
    val options: List<String>,
    val base: String,
    val activatesAutoISO: Boolean? = null,
)
